src/r1cs/linear_combination.rs

```diff
@@ -129,6 +131,53 @@ impl LinearCombination {
             self.terms.insert(var, coeff);
         }
     }
+
+    /// Extracts the non-zero weights for the left, right, output,
+    /// and witness variables, and for the constant terms, into
+    /// rows of the sparse-reduced weight matrices
+    pub fn extract_weights(
+        &self,
+    ) -> (
+        SparseWeightRow,
+        SparseWeightRow,
+        SparseWeightRow,
+        SparseWeightRow,
+        Option<Scalar>,
+    ) {
+        // Each LC can have up to `n` non-zero terms of each variable
+        // and a single constant
+
+        let mut w_l_row = SparseWeightRow(Vec::new());
+        let mut w_r_row = SparseWeightRow(Vec::new());
+        let mut w_o_row = SparseWeightRow(Vec::new());
+        let mut w_v_row = SparseWeightRow(Vec::new());
+        let mut c = None;
+
+        for (&var, &coeff) in &self.terms {
+            if coeff != Scalar::zero() {
```

Comment:
more idiomatic to use `filter` on an iterator here instead of `if coeff != Scalar::zero() { ... }`